sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -126,6 +131,10 @@ export function retryPolicy(
           }
         }
 
+        if (responseError) {
```

Comment:
Based on the below comment, I would expect this to get thrown on line 87. Is that not happening for some reason? What was happening instead?

I'm wondering if we need to rework this code to avoid the constant while condition.

Reply:
You’re correct!

At first I thought: Relying on line 87 would imply that if the retry strategies didn’t match the response we would retry up to the max retries are reached and then fail without action. Isn’t that a bug? As in, why would we retry that many times if the retry strategies couldn’t match the response?

Then I realized that this is why we have the “errorToThrow” property. While we’re retrying, we should only throw at the beginning if the error wasn’t a RestError, OR if the retry strategy sets the “errorToThrow” property.

However, to fix this on line 87 would mean to only return the response if there was no response error. This sounds ok to me at the moment, but let me know if you spot anything weird with that logic.

I’ll do the change!

Reply:
(I answered, and then I tested the change and I realized the issue)

The issue in this case is that the retry policy does not match the error nor the response in none of the strategies, and so the response goes right through the policy: no indications that we should retry or throw.

This is why we have the return on line 140: if no strategies matched this response from the service, we should return the response (and now throw the error, if there was an error).

I’ll change these messages to say something like “none of the retry strategies could work with the received response/error. Returning/throwing”

Reply:
Ah I think I see now. It makes sense if no strategy informs us to pass the result immediately.

Previously though, wouldn't it still have eventually thrown the error after the max retries was reached? Or was something preventing that?

Reply:
well, no — it would just return the response and ignore the error. I originally only tested for errors that didn’t have full HTTP responses, so I didn’t catch that. If the error didn’t have an http response, like a system error, then it would continue to loop and eventually stop, but RestErrors with responses (a rather common case), if they don’t match any strategy, they would just return the response at the end.

For the retry to happen, a strategy would need to return the “retryAfter” property or the “redirectTo” property. In this case, none is returned because the strategy has no special consideration for these network errors. 416 isn’t in any of our retry cases.

Reply:
Ahh! I see, so the issue was we checked for response not being undefined first.

I think the existing code changes are sufficient, but in the case of say maxRetry being set to 0 (for a client who doesn't want to retry), should we change

```ts
if (response !== undefined) {
  return response;
}
throw responseError;
```

into something like

```ts
if (responseError) {
   throw responseError;
} else if (response) {
  return response;
} // maybe an extra else here that throws?
```

Reply:
Done here: https://github.com/Azure/azure-sdk-for-js/pull/20883/commits/0a795e70fccc8e19d490ef6a4f7db2376ac22e0d

I also tested it with autorrest.typescript and it works like a charm 🙂 thank you for the suggestion!